### DISCUSSION TITLE
Phased backoff for non-blocking retries

### DISCUSSION THREAD

**Discussion by jgslima** [spring-projects/spring-kafka#2484] [Ideas] [answered]
Hello, I want to share some ideas and implementations I made here for the non-blocking retries. This is not a feature request, is just idea sharing.

When making retries, I think it is essential to pursue two things:

- to make the first retries quickly (with a small interval), so that for really ephemeral issues, the system recovers quickly and the user experience is less affected.
- to keep making the retries for a long time (about 4h for instance). Because some issues may take more time to go away. For example, an external system may be down, and the operations team may struggle for a while before being able to take it up again.

At first, exponential backoff seems to fit this requirement. However, to achieve the second requirement above, a large number of retry topics are needed. In a way that makes this impractical.

Fixed backoff allows a single retry topic to be used, but adopting fixed backoff brings the problem of inefficiency: in order to have quick retries and insist for a long time, one has to configure an excessive amount of attempts (with an interval of 5s, to keep trying for 4 hours, you need to configure 2880 maxAttempts, which is too much, besides too much CO2).

So I conceived here a custom `SleepingBackOffPolicy` called `PhasedSleepingBackoffPolicy`. It's objective is to have an increase in the intervals like exponential, but without an explosion of topics.
You configure for it, a list of _phases_, with which one being composed of:

- a `Duration` for the interval between retries.
- a `Duration` for the whole phase.

For example:

```yml
    phases:
    - duration: PT3M
      interval-between-retries: PT5S
    - duration: PT30M
      interval-between-retries: PT45S
    - duration: PT6H
      interval-between-retries: PT2M
```
For 3 minutes, make retries each 5 seconds, for 30 minutes, make retries each 45 seconds, for 6 hours, make retries every 2 minutes.

The amount of retry topics needed is the amount of phases.

For this to work, only the backoff policy was not enough. I had to make an extension of `DefaultDestinationTopicResolver`, which delegates part of the decision of the next retry topic to the backoff policy.  In the case of phased retries, the current `attempt` is taken into consideration when deciding in each phase (and consequently for which retry topic) the record currently is. So in terms of component design, there may be a more decoupled approach but anyway it is working fine.

**Comment by jgslima** [comment, marked as answer]
Another possibility would be for the spring-kafka to have an option to have a single retry topic, for the `maxInterval` when using exponential backoff.

For instance, for exponential with `initialInterval=1000`, `multiplier=2`, `maxInterval=8000`, and `maxAttempts`=a large number, you would have retry topics like these:

```
xxx-retry-1000
xxx-retry-2000
xxx-retry-4000
xxx-retry-8000-0
xxx-retry-8000-1
xxx-retry-8000-2
xxx-retry-8000-3
...
xxx-retry-8000-n
```

but for 8000-0 and above, it might actually be just one:

```
xxx-retry-1000
xxx-retry-2000
xxx-retry-4000
xxx-retry-8000
```
this would make it easier to adopt exponential when retrying for a long time.

**Reply by garyrussell** [reply]
@jgslima Both are good ideas; it could be argued that your second suggestion should be the default behavior - it doesn't make much sense to have multiple topics for the maxed-out delay. Unfortunately, we just missed the boat to make this the default, 3.0 was released today.

I suppose one reason for it would be to use the topic name to determine how long a record has been delayed.

cc/ @tomazfernandes Am I missing something?

With your first solution, I suppose the algorithm is such that during a phase, we keep publishing to the same topic and move to the next topic after the phase expires?

Feel free to submit a PR for either or both of these (as options).

**Reply by tomazfernandes** [reply]
> cc/ @tomazfernandes Am I missing something?

Don't think so @garyrussell, I agree both are good and viable ideas. And I'm glad the feature is extensible enough that it was possible to implement the phases on top of current logic 😊

Looking forward to the PRs @jgslima

**Reply by jgslima** [reply]
> With your first solution, I suppose the algorithm is such that during a phase, we keep publishing to the same topic and move to the next topic after the phase expires?

That's it.

> And I'm glad the feature is extensible enough that it was possible to implement the phases on top of current logic blush

Certainly, you made it very extensible, thank you and congratulations.

> Feel free to submit a PR for either or both of these (as options).

Wow.

Well, for the "phased" backoff policy, I think the custom implementation I made here is not so good in terms of component cohesion (for instance, I made my `DefaultDestinationTopicResolver` extension to be aware of, and to interact with, the backoff policy of the topic). For my custom application code I think it was fine, but for spring-kafka, I am not sure. Also my approach demanded another extension/trick, which was to call `RetryTopicConfigurationBuilder.maxAttempts()` not with the actual `maxAttempts` a record should pass through, but in the case of the phased backoff, with the number of configured phases (so that `BackOffValuesGenerator` calls the policy one for time each phase, in order to have a number of `DestinationTopic.Properties` based on the number of phases).

The second idea seems more viable to include in the framework, so I will dedicate some time here to learn how to open a PR and mainly to learn the coding and testing standards of spring-kafka.

Later, I will try to reason again about the phased backoff to see if there is a way to implement it without creating inadequate component dependencies and without artificialities.

**Reply by jgslima** [reply]
On another note, I wonder whether this "phased backoff policy" would add value to spring-retry itself, or even as an implementation of the `BackOff` interface of the spring-core module.

Personally, when configuring the policy, I think the phases configuration like I modeled above, is easier to reason about than configuring the properties of the Exponential backoff. Also, it allows you to have an initial number of retries with a small interval, instead of applying the multiplier right away (for instance, one may like to keep retrying with a very small interval for a while, say to try with 2 seconds interval for 30 seconds, and just then to increase the interval).

A possible variation of the phases configuration above might be to configure, for each phase, a `maxAttempts` for the phase, instead of a whole `Duration` for the phase. But again, personally, I like more to reason about a whole `Duration` for each phase, because it makes it easier to conclude for how long the retries will be done.

**Reply by jgslima** [reply]
Hmm, I just realized that this configuration of a Duration for the phase as a whole introduces a question, which is whether the `BackOffContext` created when calling `start(RetryContext context)` should actually take the current instant from the clock, and whether to actually keep track of how much time has passed since the beginning of each phase (my implementation just divide each phase duration by the interval of the phase, to conclude the `maxAttempts` of each phase).

If this "phased backoff" should be included in some of the frameworks, it would be better to configure directly a `maxAttempts` for each phase, to avoid this question.

**Reply by garyrussell** [reply]
It's tricky; I can see that users might want to choose both behaviors (max attempts Vs. time based).

Let's say we are in the middle of a phase and the system is taken down for extended maintenance. When the system comes back up, should the phase continue based on elapsed time since the phase started, or the attempts in this phase? I can see an argument either way.

The current mechanism ignores the downtime (the record goes to the next topic); using max attempts in each phase would take it into account and delay the movement to the next phase.

I think one or the other is fine (we don't need to complicate it further), we would just need to clearly document the behavior under this condition.

**Reply by artembilan** [reply]
Do I understand it correctly that at the moment this is possible via `RetryTemplate` and `RecoveryCallback` chain, where that `RecoveryCallback` would call the same service method via another `RetryTemplate` with its own `RetryPolicy` and `BackOffPolicy`, and so on?

**Reply by garyrussell** [reply]
`RetryTemplate` is blocking retries; this is all about non-blocking retries.

**Reply by tomazfernandes** [reply]
>  for instance, I made my DefaultDestinationTopicResolver extension to be aware of, and to interact with, the backoff policy of the topic

Just to get everyone to the same page as to how it currently works, the class that knows about `back off` values is the `DestinationTopicPropertiesFactory`, then the created `DestinationTopic.Properties` instances will be forwarded to the `DestinationTopicProcessor`, which will then populate the  map for `DestinationTopicResolver`, which has the runtime logic for resolving the topics.

So maybe we can add some property to `DestinationTopic.Properties` to be looked up by the `DTResolver` so it can know whether to keep sending to the same topic or go to the next one, or to the `DLT`.

> The current mechanism ignores the downtime (the record goes to the next topic);

Not sure it helps, but we do have a `global timeout` setting after which the message will go directly to the `DLT` - in case of an outage for example the record would timeout if so configured.

> BackOffContext created when calling start(RetryContext context) 

I'm not sure I understand how this is being used - the only time we use the `Spring Retry` interfaces is to figure out the backoff values early in the configuration phase.


I think the simplest thing to begin with would be to make it so that, depending on how the `DestinationTopic.Properties` is configured, if the current topic is the last one before the DLT, we keep sending the record to it again until a certain `maxAttempts` is reached.

We do have a `maxAttempts` property in `DestinationTopic.Properties` that currently holds the same value for all topics in a chain - maybe that's a good place to start

**Reply by jgslima** [reply]
> It's tricky; I can see that users might want to choose both behaviors (max attempts Vs. time based).

I agree. I think if we proceed at some point with this "phased" backoff policy, we should go with maxAttempts.

As you guys invited me to make some PR, initially I will create an issue to deal only with the idea of reusing a retry topic for the maxInterval delay of exponential backoffs.

**Reply by jgslima** [reply]
Created #2496
